src/middlewared/middlewared/etc_files/local/ssh/config.py

```diff
@@ -1,36 +1,84 @@
 import base64
 import os
 import re
+import stat
+import shutil
 
-from middlewared.utils import osc
+from contextlib import suppress
 
-if osc.IS_FREEBSD:
-    SSH_CONFIG_PATH = '/usr/local/etc/ssh'
-else:
-    SSH_CONFIG_PATH = '/etc/ssh'
+SSH_CONFIG_PATH = '/etc/ssh'
 
+SSH_KEYS = [
+    'ssh_host_key', 'ssh_host_key.pub', 'ssh_host_dsa_key', 'ssh_host_dsa_key.pub', 'ssh_host_dsa_key-cert.pub',
+    'ssh_host_ecdsa_key', 'ssh_host_ecdsa_key.pub', 'ssh_host_ecdsa_key-cert.pub', 'ssh_host_rsa_key',
+    'ssh_host_rsa_key.pub', 'ssh_host_rsa_key-cert.pub', 'ssh_host_ed25519_key', 'ssh_host_ed25519_key.pub',
+    'ssh_host_ed25519_key-cert.pub'
+]
 
-def generate_ssh_config(middleware):
-    ssh_config = middleware.call_sync('ssh.config')
-    for k in [
-        'ssh_host_key', 'ssh_host_key.pub', 'ssh_host_dsa_key', 'ssh_host_dsa_key.pub', 'ssh_host_dsa_key-cert.pub',
-        'ssh_host_ecdsa_key', 'ssh_host_ecdsa_key.pub', 'ssh_host_ecdsa_key-cert.pub', 'ssh_host_rsa_key',
-        'ssh_host_rsa_key.pub', 'ssh_host_rsa_key-cert.pub', 'ssh_host_ed25519_key', 'ssh_host_ed25519_key.pub',
-        'ssh_host_ed25519_key-cert.pub'
-    ]:
+DEFAULT_FILES = ['moduli', 'ssh_config', 'ssh_config.d', 'sshd_config', 'sshd_config.d']
+
+
+def generate_ssh_config(middleware, ssh_config, dirfd):
+    for k in SSH_KEYS:
+        must_remove = True
         s_key = re.sub(r'([.-])', '_', k).replace('ssh_', '', 1)
         if ssh_config[s_key]:
             decoded_key = base64.b64decode(ssh_config[s_key])
             if decoded_key:
-                with open(os.path.join(SSH_CONFIG_PATH, k), 'wb') as f:
+                mode = 0o644 if k.endswith('.pub') else 0o600
+
+                fd = os.open(k, os.O_WRONLY | os.O_CREAT, mode=mode, dir_fd=dirfd)
```

Comment:
`fd` is being leaked here?

Reply:
No. open() builtin closes the fd passed into it when exiting the context manager.

Reply:
That's nice. I didn't realize python implemented the `_at` syscalls 👍